contracts/strategies/lift/LiftStrategy.sol

```diff
@@ -0,0 +1,272 @@
+    }
+    IERC20(underlying).safeTransfer(vault, amount);
+  }
+
+  /*
+  *   Note that we currently do not have a mechanism here to include the
+  *   amount of reward that is accrued.
+  */
+  function investedUnderlyingBalance() external view returns (uint256) {
+    if (address(rewardPool) == address(0)) {
+      return IERC20(underlying).balanceOf(address(this));
+    }
+    // Adding the amount locked in the reward pool and the amount that is somehow in this contract
+    // both are in the units of "underlying"
+    // The second part is needed because there is the emergency exit mechanism
+    // which would break the assumption that all the funds are always inside of the reward pool
+    return rewardPool.balanceOf(address(this)).add(IERC20(underlying).balanceOf(address(this)));
+  }
+
+  /*
+  *   Governance or Controller can claim coins that are somehow transferred into the contract
+  *   Note that they cannot come in take away coins that are used and defined in the strategy itself
+  *   Those are protected by the "unsalvagableTokens". To check, see where those are being flagged.
+  */
+  function salvage(address recipient, address token, uint256 amount) external onlyControllerOrGovernance {
+     // To make sure that governance cannot come in and take away the coins
+    require(!unsalvagableTokens[token], "token is defined as not salvagable");
+    IERC20(token).safeTransfer(recipient, amount);
+  }
+
+  /*
+  *   Get the reward, sell it in exchange for underlying, invest what you got.
+  *   It's not much, but it's honest work.
+  *
+  *   Note that although `onlyNotPausedInvesting` is not added here,
+  *   calling `investAllUnderlying()` affectively blocks the usage of `doHardWork`
+  *   when the investing is being paused by governance.
+  */
+  function doHardWork() external onlyNotPausedInvesting restricted {
+    uint256 rewardBalanceBefore = boardRoom.getbalanceOfShare(address(this));
+    rewardPool.stakeInBoardroom();
+    uint256 rewardBalanceAfter = boardRoom.getbalanceOfShare(address(this));
+    uint256 rewardStaked = rewardBalanceAfter.sub(rewardBalanceBefore);
+    if (rewardStaked > 0) {
+      stakes.push([block.timestamp, rewardStaked]);
+    }
+    investAllUnderlying();
+  }
+
+  function _getReward() internal {
+    uint256 rewardBalanceBefore = boardRoom.getbalanceOfShare(address(this));
+    rewardPool.stakeInBoardroom();
+    uint256 rewardBalanceAfter = boardRoom.getbalanceOfShare(address(this));
+    uint256 rewardStaked = rewardBalanceAfter.sub(rewardBalanceBefore);
+    if (rewardStaked > 0) {
+      stakes.push([block.timestamp, rewardStaked]);
+    }
+  }
+
+  function withdrawRewardShareOldest() external onlyMultiSigOrGovernance {
+    require(allowedRewardClaimable, "reward claimable is not allowed");
+    uint256 len = stakes.length;
+    uint256 shareBalance = IERC20(rewardToken).balanceOf(address(this));
+    uint256 i = 0;
+    for (i;i<len;i++) {
```

Comment:
This could eventually accumulate too many stakes to be able to fit the iterations within a single block. The design requires an approach allowing for either specifying the range manually, or limiting the number of stakes of growing too high, or having a cleanup mechanism allowing to maintain a certain limit of stakes.

Reply:
Will think on this and update

Reply:
Made a solution, see latest comment